cupy/core/core.pyx

```diff
@@ -2776,21 +2776,28 @@ not_equal = create_comparison(
 
 cpdef ndarray _convert_object_with_cuda_array_interface(a):
     cdef Py_ssize_t sh, st
-    desc = a.__cuda_array_interface__
-    shape = desc['shape']
+    cdef dict desc = a.__cuda_array_interface__
```

Comment:
I think it's not safe to assume it's `dict`.
In Python you can define arbitrary type that behaves as a dictionary.

Reply:
Before I proposed the v2 protocol, I investigated and ensured all current supporting libraries (see numba/numba#4448) implemented it as a simple Python dict, so that we could safely require it to be a dict in v2 without any backward incompatibility. I believe this is a very safe and compliant change.

Reply:
Do you mean v2 already requires it to be a `dict`?
In the specification (https://numba.pydata.org/numba-doc/dev/cuda/cuda_array_interface.html), I can only find a statement that it returns a "dictionary" which is broader than `dict`.

In any case, this is the receiving side. It may be safe not to unnecessarily assume it (even if the protocol specifies it).

Reply:
> I can only find a statement that it returns a "dictionary" which is broader than dict.

That is not true. Dictionaries in Python almost always refer to the built-in `dict`. There is no ambiguity for alternative interpretation. Also, the wording and the implementation follow that of `__array_interface__` in NumPy.

I revised the wording in v2 in the hope that it can be used as a small optimization in Cython (both CuPy and mpi4py have the receiver code written in Cython), so I would very much like to keep the change as is.

Reply:
>Dictionaries in Python almost always refer to the built-in `dict`.

I can't find any (quasi-)official statement about that.
[The official glossary](https://docs.python.org/3/glossary.html#term-dictionary) describes "dictionary" as "An associative array, where arbitrary keys are mapped to values.". It's still ambiguous, I think.

Python is a duck-typing language. Without explicit wording, I wouldn't assume "dictionary" refers to `dict`.

As for performance, I did some measurement using the following code. There was no noticeable difference whether or not `d` was typed (actually it seems typed code was a bit slower, perhaps because of extra type checking).

```py
# distutils: language = c++
cpdef int func(dict d):
    cdef int s = 0
    cdef int t = 0
    for i in range(100):
        t = d['key']
        s += t
    return s
```

Also, for the purpose of optimization, an implementation of `__cuda_array_interface__` could return a specialized and faster variant of dictionary with limited set of keys, for example. Explicit typing could potentially cause problems in such a case.

Reply:
> I can't find any (quasi-)official statement about that.
> [The official glossary](https://docs.python.org/3/glossary.html#term-dictionary) describes "dictionary" as "An associative array, where arbitrary keys are mapped to values.". It's still ambiguous, I think.
> 
> Python is a duck-typing language. Without explicit wording, I wouldn't assume "dictionary" refers to `dict`.

> Also, for the purpose of optimization, an implementation of `__cuda_array_interface__` could return a specialized and faster variant of dictionary with limited set of keys, for example. Explicit typing could potentially cause problems in such a case.

I disagree. The terms "dictionary" and `dict` are always used interchangeably, see, e.g., any place in the official Python documentation that refers to dictionaries:
https://docs.python.org/3.7/library/stdtypes.html?highlight=dictionary#mapping-types-dict

In practice, having any alternative dictionary implementation for `__cuda_array_interface__` can only increase the complexity of code dependency and the risk of not being interoperable. It makes little sense to be overcautious on this.


 
> As for performance, I did some measurement using the following code. There was no noticeable difference whether or not `d` was typed (actually it seems typed code was a bit slower, perhaps because of extra type checking).
> 
> ```python
> # distutils: language = c++
> cpdef int func(dict d):
>     cdef int s = 0
>     cdef int t = 0
>     for i in range(100):
>         t = d['key']
>         s += t
>     return s
> ```
> 

This is what I got:
```python
In [1]: %load_ext Cython

In [2]: class Array:
   ...:     def __init__(self):
   ...:         self.__interface__ = {'shape': (1, 2, 3), 'ptr':18}
   ...:         

In [3]: arr = Array()

In [4]: %%cython
   ...: cpdef get_face(a):
   ...:     x = a.__interface__
   ...:     shape = x['shape']
   ...:     ptr = x['ptr']
   ...:     return ptr
   ...: 

In [5]: %%cython
    ...: cpdef get_face_typed(a):
    ...:     cdef dict x = a.__interface__
    ...:     cdef tuple shape = x['shape']
    ...:     ptr = x['ptr']
    ...:     return ptr
    ...: 

In [6]: %timeit get_face(arr)
107 ns ± 0.312 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [7]: %timeit get_face_typed(arr)
101 ns ± 0.212 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

Reply:
FYI, see the original intent of the v2 protocol for `__cuda_array_interface__` being a Python `dict`: https://github.com/numba/numba/pull/4609#issue-321035175.

Reply:
>I disagree. The terms "dictionary" and `dict` are always used interchangeably, see, e.g., any place in the official Python documentation that refers to dictionaries:
https://docs.python.org/3.7/library/stdtypes.html?highlight=dictionary#mapping-types-dict

Thank you for the pointer.
I see, maybe you're correct in that the official documentation refers to `dict` by the word "dictionary".
Maybe officially there's no ambiguity, but in reality, I think there is.
Even I, who have been using Python for 8~9 years, mistook the word, because the primary mode of the language is duck-typing and this philosophy is everywhere.

Anyway, this is about the expression in the protocol specification, and I think not essential in this discussion. As I said above, even if the protocol explicitly specifies the type to be `dict`, I don't think the receiver should assume that.

>This is what I got:

The result looks the same as mine. Looks like explicit typing has no positive impact on performance.

>In practice, having any alternative dictionary implementation for `__cuda_array_interface__` can only increase the complexity of code dependency and the risk of not being interoperable. It makes little sense to be overcautious on this.

I agree it could increase the risk of not being interoperable (not sure about code dependency). 
But it's about the producer side. Receiver side would increase the risk by making an assumption, instead.

Reply:
@niboshi, I thought you replied something but I can't find it now. Anyway, I am surprised about the pushback here, to end this and land the fix asap I changed the type in 230e83a.

Reply:
Thank you for understanding.
This is what I posted (modified a bit):

>>I disagree. The terms "dictionary" and `dict` are always used interchangeably, see, e.g., any place in the official Python documentation that refers to dictionaries:
>https://docs.python.org/3.7/library/stdtypes.html?highlight=dictionary#mapping-types-dict
>
>Thank you for the pointer.
>I see, maybe you're correct in that the official documentation refers to `dict` by the word "dictionary".
>Maybe officially there's no ambiguity, but in reality, I think there is.
>Even I, who have been using Python for 8~9 years, mistook the word, because the primary mode of the language is duck-typing and this philosophy is everywhere.
>
>Anyway, this is about the expression in the protocol specification, and I think not essential in this discussion. As I said above, even if the protocol explicitly specifies the type to be `dict`, I don't think the receiver should assume that.
>
>>This is what I got:
>
>I see. it might be dependent on environment. But anyway I think a few nanosecond is not so important.
> 
>
>>In practice, having any alternative dictionary implementation for `__cuda_array_interface__` can only increase the complexity of code dependency and the risk of not being interoperable. It makes little sense to be overcautious on this.
>
>I agree it could increase the risk of not being interoperable (not sure about code dependency). 
>But it's about the producer side. Receiver side would increase the risk by making an assumption, instead.

Reply:
Thanks, @niboshi. I have only one comment:

> > In practice, having any alternative dictionary implementation for `__cuda_array_interface__` can only increase the complexity of code dependency and the risk of not being interoperable. It makes little sense to be overcautious on this.
> 
> I agree it could increase the risk of not being interoperable (not sure about code dependency).
> But it's about the producer side. Receiver side would increase the risk by making an assumption, instead.

I still think it is a right thing to do on the receiver side to reject any non-standard `dict` alternatives. First, this is specified in the protocol as you also agreed. Second, the receiver does not (and should not) recognize the type of any non-standard dictionary, or the receiving library would have to also depend on the library that provides the custom dictionary.

Reply:
Have submitted PR ( https://github.com/numba/numba/pull/5049 ) in an attempt to clarify this in the spec. Feedback welcome 🙂